cpp/src/arrow/array/util.cc

```diff
@@ -669,11 +671,18 @@ class RepeatedArrayFactory {
   enable_if_base_binary<T, Status> Visit(const T&) {
     const std::shared_ptr<Buffer>& value = scalar<T>().value;
     std::shared_ptr<Buffer> values_buffer, offsets_buffer;
-    RETURN_NOT_OK(CreateBufferOf(value->data(), value->size(), &values_buffer));
     auto size = static_cast<typename T::offset_type>(value->size());
+
+    int64_t total_size;
```

Comment:
Should this be `typename T::offset_type`? If `total_size` is fixed to be int64_t, it always calls the int64 version of `MultiplyWithOverflow`. But `StringArray` uses int32_t for offsets so some false positive cases can go through.
For example:
```cpp
auto scalar = std::make_shared<StringScalar>("aa");
int64_t length = static_cast<int32_t>(std::numeric_limits<int32_t>::max()) / 2 + 1;
auto array_result = MakeArrayFromScalar(*scalar, length);
```
`array_result` will be `ok` but when I print the array it says `Invalid array: Negative offsets in binary array`